applications/CompressiblePotentialFlowApplication/python_scripts/define_wake_process_3d.py

```diff
@@ -128,27 +178,95 @@ def __ComputeLowerSurfaceNormals(self):
     # This function imports the stl file containing the wake and creates the wake model part out of it.
     # TODO: implement an automatic generation of the wake
     def __CreateWakeModelPart(self):
+        self.wake_model_part = self.model.CreateModelPart("wake_model_part")
+        self.dummy_property = self.wake_model_part.Properties[0]
+        self.node_id = 1
+        self.elem_id = 1
+        if not self.shed_wake_from_trailing_edge:
+            self.__ReadWakeStlModelFromFile()
+        # else: # This is now implemented in cpp
+        #     self.__ShedWakeSurfaceFromTheTrailingEdge()
+
+    def __ReadWakeStlModelFromFile(self):
+        KratosMultiphysics.Logger.PrintInfo('DefineWakeProcess3D', 'Reading wake from stl file')
         from stl import mesh #this requires numpy-stl
         wake_stl_mesh = mesh.Mesh.from_multi_file(self.wake_stl_file_name)
-        self.wake_model_part = self.model.CreateModelPart("wake_model_part")
 
-        dummy_property = self.wake_model_part.Properties[0]
-        node_id = 1
-        elem_id = 1
+        z = 0.0#-1e-4
 
         # Looping over stl meshes
         for stl_mesh in wake_stl_mesh:
             for vertex in stl_mesh.points:
-                node1 = self.wake_model_part.CreateNewNode(node_id, float(vertex[0]), float(vertex[1]), float(vertex[2]))
-                node_id+=1
-                node2 = self.wake_model_part.CreateNewNode(node_id, float(vertex[3]), float(vertex[4]), float(vertex[5]))
-                node_id+=1
-                node3 = self.wake_model_part.CreateNewNode(node_id, float(vertex[6]), float(vertex[7]), float(vertex[8]))
-                node_id+=1
+                node1 = self.__AddNodeToWakeModelPart(float(vertex[0]), float(vertex[1]), float(vertex[2]) + z )
+                node2 = self.__AddNodeToWakeModelPart(float(vertex[3]), float(vertex[4]), float(vertex[5]) + z )
+                node3 = self.__AddNodeToWakeModelPart(float(vertex[6]), float(vertex[7]), float(vertex[8]) + z )
+
+                side1 = node2 - node1
+                side2 = node3 - node1
+                face_normal = CrossProduct(side1,side2)
 
-                self.wake_model_part.CreateNewElement("Element3D3N", elem_id,  [
-                                              node1.Id, node2.Id, node3.Id], dummy_property)
-                elem_id += 1
+                normal_projection = DotProduct(face_normal, self.wake_normal)
+
+                if normal_projection >  0.0:
+                    self.__AddElementToWakeModelPart(node1.Id, node2.Id, node3.Id)
+                else:
+                    self.__AddElementToWakeModelPart(node1.Id, node3.Id, node2.Id)
+
+    def __ShedWakeSurfaceFromTheTrailingEdge(self):
+        KratosMultiphysics.Logger.PrintInfo('DefineWakeProcess3D', 'Shedding wake from trailing edge')
+        number_of_elements_in_wake_direction = int(self.shedded_wake_distance/self.shedded_wake_element_size)
+        z = 1e-9
+
```

Comment:
Do you want to keep the python code? I would leave the cpp only

Reply:
Honestly I personally find the python convenient when developing something new to try things fast, but at the same time it is code duplication (or even worse, python and cpp do different things like in this case). So I don't know what is better, we can remove it if you want.

Reply:
I think at the stage we are at we can just stick to the cpp and keep the python clean. If you want to keep the python code we can store it for some time in a branch.

Reply:
makes sense :+1: , Ill remove it then